### ISSUE TITLE
Vaadin Flow 24.4.0.alpha4 prod build  - Caused by: java.lang.NoClassDefFoundError: com/vaadin/base/devserver/DevToolsMessageHandler

### ISSUE THREAD

**Issue by alexanoid** [vaadin/flow#18734]
### Description of the bug

I upgraded from 24.4.0.alpha2 to 24.4.0.alpha4, created the prod build and now the application is failed to start with the following exception:

```
org.springframework.context.ApplicationContextException: Failed to start bean 'webServerStartStop'
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:287) ~[spring-context-6.1.3.jar!/:6.1.3]
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:467) ~[spring-context-6.1.3.jar!/:6.1.3]
	at java.base/java.lang.Iterable.forEach(Iterable.java:75) ~[na:na]
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:256) ~[spring-context-6.1.3.jar!/:6.1.3]
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:201) ~[spring-context-6.1.3.jar!/:6.1.3]
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:978) ~[spring-context-6.1.3.jar!/:6.1.3]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:627) ~[spring-context-6.1.3.jar!/:6.1.3]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:334) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1354) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1343) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at com.example.ui.Application.main(Application.java:31) ~[!/:0.0.1]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:91) ~[ui-0.0.1.jar:0.0.1]
	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:53) ~[ui-0.0.1.jar:0.0.1]
	at org.springframework.boot.loader.launch.JarLauncher.main(JarLauncher.java:58) ~[ui-0.0.1.jar:0.0.1]
Caused by: java.lang.NoClassDefFoundError: com/vaadin/base/devserver/DevToolsMessageHandler
	at java.base/java.lang.ClassLoader.defineClass1(Native Method) ~[na:na]
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017) ~[na:na]
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150) ~[na:na]
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:524) ~[na:na]
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:427) ~[na:na]
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:421) ~[na:na]
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712) ~[na:na]
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:420) ~[na:na]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:592) ~[na:na]
	at org.springframework.boot.loader.net.protocol.jar.JarUrlClassLoader.loadClass(JarUrlClassLoader.java:104) ~[ui-0.0.1.jar:0.0.1]
	at org.springframework.boot.loader.launch.LaunchedClassLoader.loadClass(LaunchedClassLoader.java:91) ~[ui-0.0.1.jar:0.0.1]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[na:na]
	at com.vaadin.copilot.CopilotIndexHtmlLoader.serviceInit(CopilotIndexHtmlLoader.java:16) ~[copilot-24.4.0.alpha2.jar!/:na]
	at com.vaadin.flow.server.VaadinService.lambda$init$0(VaadinService.java:235) ~[flow-server-24.4.0.alpha7.jar!/:24.4.0.alpha7]
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133) ~[na:na]
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845) ~[na:na]
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762) ~[na:na]
	at com.vaadin.flow.server.VaadinService.lambda$init$1(VaadinService.java:235) ~[flow-server-24.4.0.alpha7.jar!/:24.4.0.alpha7]
	at com.vaadin.flow.server.VaadinService.runWithServiceContext(VaadinService.java:2369) ~[flow-server-24.4.0.alpha7.jar!/:24.4.0.alpha7]
	at com.vaadin.flow.server.VaadinService.init(VaadinService.java:233) ~[flow-server-24.4.0.alpha7.jar!/:24.4.0.alpha7]
	at com.vaadin.flow.spring.SpringVaadinServletService.init(SpringVaadinServletService.java:102) ~[vaadin-spring-24.4.0.alpha7.jar!/:na]
	at com.vaadin.flow.spring.SpringServlet.createServletService(SpringServlet.java:115) ~[vaadin-spring-24.4.0.alpha7.jar!/:na]
	at com.vaadin.flow.server.VaadinServlet.createServletService(VaadinServlet.java:336) ~[flow-server-24.4.0.alpha7.jar!/:24.4.0.alpha7]
	at com.vaadin.flow.server.VaadinServlet.init(VaadinServlet.java:132) ~[flow-server-24.4.0.alpha7.jar!/:24.4.0.alpha7]
	at io.undertow.servlet.core.LifecyleInterceptorInvocation.proceed(LifecyleInterceptorInvocation.java:117) ~[undertow-servlet-2.3.10.Final.jar!/:2.3.10.Final]
	at io.undertow.servlet.core.ManagedServlet$DefaultInstanceStrategy.start(ManagedServlet.java:304) ~[undertow-servlet-2.3.10.Final.jar!/:2.3.10.Final]
	at io.undertow.servlet.core.ManagedServlet.createServlet(ManagedServlet.java:140) ~[undertow-servlet-2.3.10.Final.jar!/:2.3.10.Final]
	at io.undertow.servlet.core.DeploymentManagerImpl$2.call(DeploymentManagerImpl.java:589) ~[undertow-servlet-2.3.10.Final.jar!/:2.3.10.Final]
	at io.undertow.servlet.core.DeploymentManagerImpl$2.call(DeploymentManagerImpl.java:559) ~[undertow-servlet-2.3.10.Final.jar!/:2.3.10.Final]
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:42) ~[undertow-servlet-2.3.10.Final.jar!/:2.3.10.Final]
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43) ~[undertow-servlet-2.3.10.Final.jar!/:2.3.10.Final]
	at io.undertow.servlet.core.DeploymentManagerImpl.start(DeploymentManagerImpl.java:605) ~[undertow-servlet-2.3.10.Final.jar!/:2.3.10.Final]
	at org.springframework.boot.web.embedded.undertow.DeploymentManagerHttpHandlerFactory$DeploymentManagerHandler.<init>(DeploymentManagerHttpHandlerFactory.java:65) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.web.embedded.undertow.DeploymentManagerHttpHandlerFactory.getHandler(DeploymentManagerHttpHandlerFactory.java:46) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.web.embedded.undertow.UndertowWebServer.createHttpHandler(UndertowWebServer.java:172) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.web.embedded.undertow.UndertowServletWebServer.createHttpHandler(UndertowServletWebServer.java:72) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.web.embedded.undertow.UndertowWebServer.createUndertowServer(UndertowWebServer.java:164) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.web.embedded.undertow.UndertowWebServer.start(UndertowWebServer.java:116) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.boot.web.servlet.context.WebServerStartStopLifecycle.start(WebServerStartStopLifecycle.java:44) ~[spring-boot-3.2.2.jar!/:3.2.2]
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:284) ~[spring-context-6.1.3.jar!/:6.1.3]
	... 20 common frames omitted
Caused by: java.lang.ClassNotFoundException: com.vaadin.base.devserver.DevToolsMessageHandler
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445) ~[na:na]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:592) ~[na:na]
	at org.springframework.boot.loader.net.protocol.jar.JarUrlClassLoader.loadClass(JarUrlClassLoader.java:104) ~[ui-0.0.1.jar:0.0.1]
	at org.springframework.boot.loader.launch.LaunchedClassLoader.loadClass(LaunchedClassLoader.java:91) ~[ui-0.0.1.jar:0.0.1]
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[na:na]
	... 61 common frames omitted
```

### Expected behavior

Application should start without any errors

### Minimal reproducible example

n/a

### Versions

- Vaadin / Flow version: 24.4.0.alpha4

**Comment by knoobie** [comment]
Doesn't look like a proper production build while `CopilotIndexHtmlLoader` is in the class path.

**Comment by alexanoid** [comment]
I simply updated the version from 24.4.0.alpha2 to 24.4.0.alpha4. I didn't change any configurations. Everything works fine on 24.4.0.alpha2, but on 24.4.0.alpha4, it crashes with an error.

This is my Maven prod profile config:

```
<profile>
            <id>prod</id>
            <properties>
                <activatedProperties>prod</activatedProperties>
                <vaadin.productionMode>true</vaadin.productionMode>
            </properties>

            <build>
                <plugins>

                    <plugin>
                        <groupId>com.vaadin</groupId>
                        <artifactId>vaadin-maven-plugin</artifactId>
                        <version>${vaadin.version}</version>
                        <executions>
                            <execution>
                                <goals>
                                    <goal>build-frontend</goal>
                                </goals>
                                <phase>compile</phase>
                            </execution>
                        </executions>
                    </plugin>

                    <plugin>
                        <groupId>org.apache.maven.plugins</groupId>
                        <artifactId>maven-surefire-plugin</artifactId>
                        <configuration>
                            <skipTests>true</skipTests>
                        </configuration>
                    </plugin>

                </plugins>
            </build>

            <dependencies>
                <dependency>
                    <groupId>com.vaadin</groupId>
                    <artifactId>vaadin</artifactId>
                    <exclusions>
                        <exclusion>
                            <groupId>com.vaadin</groupId>
                            <artifactId>vaadin-dev-server</artifactId>
                        </exclusion>
                    </exclusions>
                </dependency>
            </dependencies>

        </profile>
```

**Comment by knoobie** [comment]
You are excluding the wrong dependency. It's vaadin-dev. 

Keep in mind it's an alpha with a lot of things still on going and no migration path documented.

**Comment by alexanoid** [comment]
> You are excluding the wrong dependency. It's vaadin-dev.
> 
> Keep in mind it's an alpha with a lot of things still on going and no migration path documented.

Thank you! I added the following:

```
            <dependencies>
                <dependency>
                    <groupId>com.vaadin</groupId>
                    <artifactId>vaadin</artifactId>
                    <exclusions>
                        <exclusion>
                            <groupId>com.vaadin</groupId>
                            <artifactId>vaadin-dev</artifactId>
                        </exclusion>
                        <exclusion>
                            <groupId>com.vaadin</groupId>
                            <artifactId>vaadin-dev-bundle</artifactId>
                        </exclusion>
                        <exclusion>
                            <groupId>com.vaadin</groupId>
                            <artifactId>vaadin-dev-server</artifactId>
                        </exclusion>
                    </exclusions>
                </dependency>
            </dependencies>
```

and now everything is working fine!

**Comment by alexanoid** [comment]
@knoobie One more question: With 24.4.0.alpha4, something changed in the form validation process. Once I enter the first value, it immediately highlights the entire form with validation errors on the mandatory fields:

![form](https://github.com/vaadin/flow/assets/110009335/bafe1f1a-5f0e-4eae-827a-613667293539)

Previously, in version 24.4.0.alpha2, the validation started working only after I pressed the Save button. How can I control this behavior in version 24.4.0.alpha4?

**Comment by knoobie** [comment]
I don't think this is intended, sounds like a bug to me. I would suggest to create a new issue for that. 

CC: @mshabarov this looks severe

**Comment by alexanoid** [comment]
Thanks, created new issue https://github.com/vaadin/flow/issues/18735 for this case and for one another with RouterLink in the Drawer on mobile view

**Comment by mshabarov** [comment]
> Thank you! I added the following:

Excluding only
```
<exclusion>
   <groupId>com.vaadin</groupId>
   <artifactId>vaadin-dev</artifactId>
</exclusion>
```
should be enough, if it's not enough for your setup and produces issues, please create an issue in Flow.


> @knoobie One more question: With 24.4.0.alpha4, something changed in the form validation process. Once I enter the first value, it immediately highlights the entire form with validation errors on the mandatory fields:
> 
> Previously, in version 24.4.0.alpha2, the validation started working only after I pressed the Save button. How can I control this behavior in version 24.4.0.alpha4?

Doesn't look proper to me, we'll take a look, thanks!
